### ISSUE TITLE
[GoRouter] ShellRoute state does not take pushed route into account.

### ISSUE THREAD

**Issue by cgestes** [flutter/flutter#114901]
```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() => runApp(App());

final GoRouter _router = GoRouter(
  routes: <RouteBase>[
    ShellRoute(
        builder: (context, state, child) {
          print("Shell: ${state.location}");
          return child;
        },
        routes: <GoRoute>[
          GoRoute(
            path: '/',
            builder: (BuildContext context, GoRouterState state) {
              return MyHomePage();
            },
          ),
          GoRoute(
            path: '/b',
            builder: (BuildContext context, GoRouterState state) {
              return MyHomePageB();
            },
          ),
        ])
  ],
);

class App extends StatelessWidget {
  App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerConfig: _router,
      title: 'GoRouter Example',
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    final st = GoRouterState.of(context);
    print("locA: ${st.location}");
    return Scaffold(
        appBar: AppBar(
          title: const Text("A"),
        ),
        body: Column(children: [
          TextButton(child: Text("go A"), onPressed: () => _router.go('/')),
          TextButton(child: Text("push A"), onPressed: () => _router.push('/')),
          TextButton(child: Text("go B"), onPressed: () => _router.go('/b')),
          TextButton(
              child: Text("push B"), onPressed: () => _router.push('/b')),
        ]));
  }
}

class MyHomePageB extends StatelessWidget {
  const MyHomePageB({super.key});

  @override
  Widget build(BuildContext context) {
    final st = GoRouterState.of(context);
    print("locB: ${st.location}");
    return Scaffold(
        appBar: AppBar(
          title: const Text("B"),
        ),
        body: Column(children: [
          TextButton(child: Text("go A"), onPressed: () => _router.go('/')),
          TextButton(child: Text("push A"), onPressed: () => _router.push('/')),
          TextButton(child: Text("go B"), onPressed: () => _router.go('/b')),
          TextButton(
              child: Text("push B"), onPressed: () => _router.push('/b')),
        ]));
  }
}
```

Launch, click on pushA or pushB, observe that the shell state doesn't take the pushed route into account.

**Comment by cgestes** [comment]
If this is the expected behavior, how do I get the state of the last pushed route?

**Comment by cgestes** [comment]
feel free to close if this is the expected behavior.

**Comment by cgestes** [comment]
I did implement my own navigation history and stopped using push.

(grabbing the history was also a bit challenge, the only official way I found was to rely on NavObserver which emits multiple operations for a single router.go() and generally does not map well to go router).

So I did subscribe to a few "addListener" (not sure how correct this is):
```dart
    router.routeInformationProvider.addListener(onChange);
    router.routerDelegate.addListener(onChange);
    router.addListener(onChange);
```

So I guess I'll close, as I dont have issue anymore. I can open separate bugs for other parts I mention if needed.

Thanks for this great package.
